### PR TITLE
Feature/embed external exercises

### PR DESCRIPTION
**Pull request by pandafulmanda**
Unsure if i should be doing a PR?  Seems like a good idea.

will create separate branch for body.coffee renderDom refactor

### CODE REVIEW

File: src/data/exercises.json [hunk 1]
```diff
@@ -0,0 +1,455 @@
+{
+    "total_count": 10,
```
Comment (philschatz) [inline src/data/exercises.json]: I believe this JSON is for a set of exercises but the `<a href="...">` in the content HTML points to a single exercise. Thoughts @Dantemss?

Comment (Dantemss) [inline src/data/exercises.json]: Correct, this is the result set with 10 exercises.

For an example of what you will get for searching for 1 embed tag, see here: https://exercises-dev1.openstax.org/api/exercises?q=tag:k12phys-ch04-ex001

Searching for multiple tags at once:
https://exercises-dev1.openstax.org/api/exercises?q=tag:k12phys-ch04-ex001,k12phys-ch04-ex002

And if you search for a tag that's in a bunch of exercises, like a LO, you will get a lot of results: https://exercises-dev1.openstax.org/api/exercises?q=tag:k12phys-ch04-s01-lo01

Comment (Dantemss) [inline src/data/exercises.json]: These URL's might need to change slightly once exercises is public (we don't want exercises from random people, only trusted authors or exercises in certain lists), but for now they should work :)

Comment (pandafulmanda) [inline src/data/exercises.json]: The JSON is a stub for the real thing that would return only the one 1 embed tag that is relevant.  I could put extra logic in to sort the "results" for items that match only the tag, but that would be unnecessary when using the real API.

I guess I put something in to handle both cases

File: src/scripts/configs/embeddables.coffee [hunk 2]
```diff
@@ -0,0 +1,41 @@
+define (require) ->
+
+  _ = require('underscore')
```
Comment (dak) [inline src/scripts/configs/embeddables.coffee]: It doesn't appear that underscore is a dependency of this file (the only place I see it being used is commented out).

File: src/scripts/configs/embeddables.coffee [hunk 3]
```diff
@@ -0,0 +1,41 @@
...
+    },{
+      match : '#ost\/api\/ex\/'
+      matchType : 'a'
+      embeddableType : 'exercise'
+      apiUrl : () ->
+        # # stub.  Comment out to use local exercises stub.
+        # # Copied from https://exercises-dev1.openstax.org/api/exercises?q=tag:k12phys-ch04-s01-lo01
+        # 'http://localhost:8000/data/exercises.json';
+
+        'https://exercises-dev1.openstax.org/api/exercises?q=tag:' + @.itemCode;
```
Comment (dak) [inline src/scripts/configs/embeddables.coffee]: Should this be configured in `settings.js`?

File: src/scripts/configs/embeddables.coffee [hunk 4]
```diff
@@ -0,0 +1,41 @@
...
+      #     _.indexOf(item.tags, @itemCode) > -1
+      #   , @)
+
+      onRender : ($parent) ->
+        # For the maths.  you know. it do what it do.
+        $mathElements = $parent.find('[data-math]')
+        $mathElements.each((iter, element) ->
+          mathTex = '[TEX_START]' + $(element).data('math') + '[TEX_END]'
+          $(element).text(mathTex)
+        )
```
Comment (dak) [inline src/scripts/configs/embeddables.coffee]: We've been avoiding wrapping multi-line params in parens as a general style.  See below:

```coffeescript
$mathElements.each (iter, element) ->
  mathTex = '[TEX_START]' + $(element).data('math') + '[TEX_END]'
  $(element).text(mathTex)
```

File: src/scripts/modules/media/body/body.coffee [hunk 5]
```diff
@@ -1,11 +1,24 @@
...
   Mathjax = require('mathjax')
   router = require('cs!router')
   EditableView = require('cs!helpers/backbone/views/editable')
   template = require('hbs!./body-template')
   require('less!./body')
 
+  embeddableTemplates = {
+    'exercise' : require('hbs!./embeddables/exercise-template'),
+    'iframe' : require('hbs!./embeddables/iframe-template')
+  }
```
Comment (dak) [inline src/scripts/modules/media/body/body.coffee]: As a general style we've been dropping multi-line brackets and commas when possible:
```coffeescript
embeddableTemplates =
  'exercise' : require('hbs!./embeddables/exercise-template')
  'iframe' : require('hbs!./embeddables/iframe-template')
```

File: src/scripts/modules/media/body/body.coffee [hunk 6]
```diff
@@ -125,12 +134,201 @@ define (require) ->
...
+      # finds fresh element in @$el if a selector is provided
+      # instead of the element itself.
+
+      embeddableItem.$el = @$el.find(embeddableItem.selector) if embeddableItem.selector
+
+      $parent = embeddableItem.$el.parent()
+      embeddableItem.$el.replaceWith(embeddableItem.html)
+
+      if _.isFunction(embeddableItem.onRender)
+        embeddableItem.onRender($parent)
```
Comment (dak) [inline src/scripts/modules/media/body/body.coffee]: Coffeescript offers an easier way to check if something is a function:
```coffeescript
embeddableItem.onRender?($parent)
```

Compiles to:
```
if (typeof embeddableItem.onRender === "function") {
  embeddableItem.onRender($parent);
}
```

File: src/scripts/modules/media/body/body.coffee [hunk 7]
```diff
@@ -125,12 +134,201 @@ define (require) ->
...
+    # Could abstract a little more and pull out to another file if needed.
+    ###
+
+    # A queue for rendering embeddables means
+    # we can share rendering functions regardless
+    # of whether the embeddable is sync or async
+    initializeEmbeddableQueues: () ->
+      @renderEmbeddableQueue = []
+
+      _.extend(@renderEmbeddableQueue, Backbone.Events)
```
Comment (dak) [inline src/scripts/modules/media/body/body.coffee]: I haven't looked at this closely to comment on it intelligently, but as a quick glance this strikes me as something that needs a closer look as possibly being overkill.

**Comment by dak** [comment]
@pandafulmanda It looks like the build is failing because `settings.js` is not beautified (some formatting issue probably).  Also, can you try squashing some of these commits?

File: src/data/exercises.json [hunk 8]
```diff
@@ -0,0 +1,1040 @@
...
+          "formats": [
+            "multiple-choice",
+            "free-response"
+          ],
+          "combo_choices": []
+        }
+      ]
+    }
+  ]
+}
```
Comment (dak) [inline src/data/exercises.json]: Can you run this through `http://jsbeautifier.org` (with 2 spaces option)?

Also, for largely [historical reasons](http://stackoverflow.com/questions/2287967/why-is-it-recommended-to-have-empty-line-in-the-end-of-file), all files in webview should end in a new-line (ie, have a blank last line)

File: src/scripts/configs/embeddables.coffee [hunk 9]
```diff
@@ -0,0 +1,44 @@
+define (require) ->
+
+  settings = require('settings')
+
+  return {
+    embeddableTypes : [{
+      match : '#terp-'
+      matchType : 'a'
+      embeddableType : 'terp'
+      apiUrl : () ->
```
Comment (dak) [inline src/scripts/configs/embeddables.coffee]: There's a bunch of files in this commit with a space between the property and colon.  You may want to do a search replace for ` : ` -> `: `.

**Pull request merged**